text/0580-destroyables.md

```diff
@@ -0,0 +1,163 @@
+   historically what led to the actions hash on classic components, in order to
+   avoid collisions between actions named `destroy` and the `destroy` lifecyle
+   hook.
+2. On a related note, relying on property names means that all framework classes
+   _must_ implement the `willDestroy` function (or another name), making it very
+   difficult to change APIs in the future.
+3. Methods are difficult for _libraries_ to instrument. For instance,
+   `ember-concurrency` currently replaces the `willDestroy` method on any class
+   with a task, with logic that looks similar to:
+
+   ```js
+   let PATCHED = new WeakSet();
+
+   function patchWillDestroy(obj) {
+     if (PATCHED.has(obj)) return;
+
+     let oldWillDestroy = obj.willDestroy;
+
+     obj.willDestroy = function() {
+       if (oldWillDestroy) oldWillDestroy.call(this);
+
+       teardownTasks(this);
+     }
+
+     PATCHED.add(obj);
+   }
+   ```
+
+   This logic becomes especially convoluted if _multiple_ libraries are
+   attempting to patch `willDestroy` in this way.
+
+4. Finally, since this isn't a standard, it's difficult to add _layers_ of new
+   destroyable values that can interoperate with one another. For instance,
+   there is no way for `ember-concurrency` to know how to destroy tasks on
+   non-framework classes that users may have added themselves.
+
+This RFC proposes a streamlined API that disconnects the exact implementation
+from any interface, allows for multiple destructors per-destroyable, and
+maximizes interoperability in general.
+
+## Detailed design
+
+The API consists of 3 functions:
+
+```ts
+declare function registerLifetime(parent: object, child: object): void;
+declare function registerDestructor(destroyable: object, destructor: () => void): void;
+declare function destroy(destroyable: object): void;
+```
+
+### `registerLifetime`
+
+This function is used to associate a destroyable object's lifetime with a
+parent. When the parent is destroyed, all registered children will also be
+destroyed.
+
+### `registerDestructor`
+
+Receives a destroyable and a destructor function, and associates the function
+with it. When the destroyable is destroyed, or when its parent is destroyed, the
+destructor function will be called. Multiple destructors can be associated with
+a given destroyable, and they can be associated over time, allowing libraries
+like `ember-lifeline` to dynamically add destructors as needed.
+
+### `destroy`
+
+`destroy` manually initiates the destruction of a destroyable. This allows users
+to manually manage the lifecycles of destroyables that are shorter lived than
+their parents (for instance, destroyables that are defined an a service, which
+lives as long as an application does), cleaning them up as needed.
+
+### Built In Destroyables
+
+The root destroyable of an Ember application will be the instance of the
+application itself. All framework managed classes are destroyables, including:
```

Comment:
> will be the instance of the application itself

technically it is the instance of Ember.ApplicationInstance (not an "instance of the application itself")

Maybe you could rephrase to "owner"?